data/tutorials/language/0lg_09_env.md

```diff
@@ -0,0 +1,599 @@
+- No global definition is introduced in this example
+- Computation of `2 * 3` will always take place before `b * 7`
+
+Local definitions can be chained (one after another) or nested (one inside another). Here is an example of chaining:
+```ocaml
+# let b = 2 * 3 in
+  let c = b * 7 in
+  b * c;;
+- : int = 252
+```
+
+This is how scoping works here:
+- `b` is bound to `6` inside `let c = b * 7 in b * c`
+- `c` is bound to `42` inside `b * c`
+
+Here is an example of nesting:
+```ocaml
+# let b =
+    let c = 2 * 3 in
+    c * 5 in
+  b * 7;;
+- : int = 210
+```
+Here is how scoping works:
+- `c` is bound to `6` inside `c * 5`
+- `b` is bound to `30` inside `b * 7`
+
+Arbitrary combinations of chaining or nesting are allowed.
+
+## Scopes and Environments
+
+Without oversimplifying, an OCaml program is a sequence of definitions or expressions. These items are said to be at the _top level_. An OCaml REPL, such as UTop, is called a toplevel because that's where typed expressions go.
+
+Execution evaluates each expression from top to bottom.
+
+At any time during evaluation, the _environment_ is the ordered sequence of available definitions.
+
+Here, the name `tau` is added into the toplevel environment.
+```ocaml
+# let tau = 6.28318;;
+val tau : float = 6.28318
+```
+
+The scope of `tau` is global. This name is available anywhere after the definition.
+
+Here, the global environment is unchanged:
+```ocaml
+# let pi = 3.14159 in 2. *. pi;;
+- : float = 6.28318
+```
+
+However, with respect to the expression `2. *. pi`, the environment is different because it contains the definition of `pi`. Local definitions create local environments.
+
+The scope of a definition is the set of environments where it is reachable.
+
+Although OCaml is an expression-oriented language, it is not entirely free of statements. The `let` construct is a statement that may change the state of the environment by adding a name-value binding. In some sense, the `let` is the only statement OCaml has. Note these expressions at the toplevel also fall into that category because they are equivalent to `let _ =` definitions.
+```ocaml
+# (1.0 +. sqrt 5.0) /. 2.0;;
+- : float = 1.6180339887498949
+
+# let _ = (1.0 +. sqrt 5.0) /. 2.0;;
+- : float = 1.6180339887498949
+```
+
+With respect to the environment, there are no means to:
+- List its contents
+- Remove a definition
+- Reset it to an earlier state
+
+### Same-Level Shadowing
+
+You can write two definitions using the same identifier in sequence:
+```ocaml
+# let c = 2 * 3;;
+val c : int = 6
+
+# let d = c * 7;;
+val d : int = 42
+
+# let c = 7 * 7;;
+val c : int = 49
+
+# d;;
+- : int = 42
+```
+
+In this example, `c` is defined twice. The key thing to understand is that `c` is not updated. It looks as if the value of `c` has changed, but it hasn't. When the second definition takes place, the first one becomes unreachable, but it still remains in the environment.
```

Comment:
Does this mean you can change the definition for one expression, but after it's processed, the definition goes back to the original definition (because it's immutable)?